File: backend/app/schemas/options.py
from pydantic import BaseModel


# ──────────────────────────────────────────────
# Greeks
# ──────────────────────────────────────────────


class GreeksOut(BaseModel):
    price: float
    delta: float
    gamma: float
    theta: float
    vega: float
    rho: float


# ──────────────────────────────────────────────
# IV / Skew / Term Structure
# ──────────────────────────────────────────────


class IVMetricsOut(BaseModel):
    current_iv: float
    iv_rank: float
    iv_percentile: float
    iv_high: float
    iv_low: float


class SkewPointOut(BaseModel):
    strike: float
    call_iv: float | None
    put_iv: float | None


class SkewOut(BaseModel):
    skew_ratio: float
    points: list[SkewPointOut]


class TermStructurePointOut(BaseModel):
    expiration: str
    days_to_expiry: int
    atm_iv: float


class OptionsAnalysisOut(BaseModel):
    symbol: str
    spot_price: float
    expirations: list[str]
    iv_metrics: IVMetricsOut
    skew: SkewOut
    term_structure: list[TermStructurePointOut]


# ──────────────────────────────────────────────
# LEAPS
# ──────────────────────────────────────────────


class ThetaEfficiencyOut(BaseModel):
    delta_per_dollar: float
    theta_per_delta: float


class LeapsCandidateOut(BaseModel):
    strike: float
    expiration: str
    option_type: str
    days_to_expiry: int
    market_price: float
    delta: float
    theta: float
    vega: float
    iv: float
    intrinsic: float
    extrinsic: float
    extrinsic_pct: float
    theta_efficiency: ThetaEfficiencyOut
    stock_replacement_cost: float
    roll_recommendation: str


class LeapsAnalysisOut(BaseModel):
    symbol: str
    spot_price: float
    leaps_expirations: list[str]
    candidates: list[LeapsCandidateOut]

File: frontend/src/types/options.ts
export interface IVMetrics {
  current_iv: number;
  iv_rank: number;
  iv_percentile: number;
  iv_high: number;
  iv_low: number;
}

export interface SkewPoint {
  strike: number;
  call_iv: number | null;
  put_iv: number | null;
}

export interface Skew {
  skew_ratio: number;
  points: SkewPoint[];
}

export interface TermStructurePoint {
  expiration: string;
  days_to_expiry: number;
  atm_iv: number;
}

export interface OptionsAnalysis {
  symbol: string;
  spot_price: number;
  expirations: string[];
  iv_metrics: IVMetrics;
  skew: Skew;
  term_structure: TermStructurePoint[];
}

export interface ThetaEfficiency {
  delta_per_dollar: number;
  theta_per_delta: number;
}

export interface LeapsCandidate {
  strike: number;
  expiration: string;
  option_type: string;
  days_to_expiry: number;
  market_price: number;
  delta: number;
  theta: number;
  vega: number;
  iv: number;
  intrinsic: number;
  extrinsic: number;
  extrinsic_pct: number;
  theta_efficiency: ThetaEfficiency;
  stock_replacement_cost: number;
  roll_recommendation: string;
}

export interface LeapsAnalysis {
  symbol: string;
  spot_price: number;
  leaps_expirations: string[];
  candidates: LeapsCandidate[];
}
